### PR TITLE
Added redux to project

### PR DESCRIPTION
**Pull request by nathan-hyan**
## Summary

Added redux, actions and reducers

## Screenshots

[Change!] Show the screenshots of the views you modified.

### Android

image

### iOS

image

## Trello Card

[Card 1](https://trello.com/c/7UUIhQV1)

[Card 2](https://trello.com/c/bBQqvAVv)

### CODE REVIEW

File: wbooks/src/constants/types.ts [hunk 1]
```diff
@@ -0,0 +1 @@
+export const BOOKS = { GET: 'GET_BOOKS' };
```
Comment (sfernandez11) [inline wbooks/src/constants/types.ts:1]: Why this? You already have the `actions` object 🤔

File: wbooks/src/redux/book/reducer.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import { BOOKS } from '@constants/types';
```
Comment (sfernandez11) [inline wbooks/src/redux/book/reducer.ts:1]: No, use the actions object from `./actions`

File: wbooks/src/redux/book/reducer.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+import { actionCreators } from './actions';
```
Comment (sfernandez11) [inline wbooks/src/redux/book/reducer.ts:3]: This is not correct, the actionCreators must be called from another actions or components/screens. Never put logic in the reducer, it is only for state managment

File: wbooks/src/redux/book/reducer.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+const initialState = {
+  ok: false,
+  data: []
+};
```
Comment (sfernandez11) [inline wbooks/src/redux/book/reducer.ts:8]: No, the initialState must be:
```
const initialState = {
 books: [],
 booksLoading: false,
 bookError: null
}
```
`ok`  and `data` are the `apisauce` or service response, not what you save in redux

File: wbooks/src/redux/book/reducer.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+interface Action {
+  type: string;
```
Comment (sfernandez11) [inline wbooks/src/redux/book/reducer.ts:11]: You must add the payload here too

File: wbooks/src/redux/book/reducer.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+}
+
+export const bookReducer = (state = initialState, action: Action) => {
+  switch (action.type) {
+    case BOOKS.GET:
```
Comment (sfernandez11) [inline wbooks/src/redux/book/reducer.ts:16]: `case actions.GET_BOOK:`

File: wbooks/src/redux/book/reducer.ts [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+      return {
+        ok: true,
+        data: actionCreators.getBooks()
+      };
```
Comment (sfernandez11) [inline wbooks/src/redux/book/reducer.ts:20]: No, in the `GET_BOOK` you must set `booksLoading: true`
In `GET_BOOKS_SUCCESS` set `booksLoading: false` and `books: action.payload` and you forgot `GET_BOOK_FAILURE` and save `booksError: action.payload` 

Remember that state is an object

File: wbooks/src/redux/store.ts [hunk 8]
```diff
@@ -0,0 +1,9 @@
+import { applyMiddleware, createStore, compose, combineReducers } from 'redux';
+import thunk from 'redux-thunk';
+import Reactotron from '@config/reactotronConfig';
+
+import { bookReducer } from './book/reducer';
+
+const rootReducer = combineReducers({ bookReducer });
+
+export default createStore(rootReducer, compose(applyMiddleware(thunk), Reactotron.createEnhancer()));
```
Comment (sfernandez11) [inline wbooks/src/redux/store.ts:9]: Be careful with this, Reactotron only be in `__DEV__` mode:
```
const rootReducer = combineReducers({
  bookReducer
});

const middlewares = [applyMiddleware(thunk)];

if (__DEV__) {
  middlewares.push(Reactotron.createEnhancer());
}

const store = createStore(rootReducer, compose(...middlewares));

export default store;
```

**Review by sfernandez11 — CHANGES_REQUESTED** [review]
Great, some comments...

File: wbooks/App.test.tsx [hunk 9]
```diff
@@ -4,13 +4,25 @@
...
 import React from 'react';
+import { Provider } from 'react-redux';
+import configureMockStore from 'redux-mock-store';
+import thunk from 'redux-thunk';
 import renderer from 'react-test-renderer';
 
 import App from './src/app';
 
-// Note: test renderer must be required after react-native.
+const mockStore = configureMockStore([thunk]);
```
Comment (sfernandez11) [inline wbooks/App.test.tsx:14]: Nice, we can implement this in osde 😄

File: wbooks/src/app/screens/BookList/index.tsx [hunk 10]
```diff
@@ -1,18 +1,32 @@
...
+import actionCreators from '@redux/book/actions';
 
 import BookItem from './components/BookItem';
 import styles from './styles';
 
+interface Props {
+  bookReducer: {
+    books: Book[];
+  };
+}
```
Comment (sfernandez11) [inline wbooks/src/app/screens/BookList/index.tsx:14]: Why this? You must define an AppState interface (in another file) for this:
```
interface AppState {
 book: BookState;
}
```
And define BookState with the book reducer state interface

File: wbooks/src/app/screens/BookList/index.tsx [hunk 11]
```diff
@@ -1,18 +1,32 @@
...
+
 function BookList() {
+  const dispatch = useDispatch();
+  const books = useSelector((state: Props) => state.bookReducer.books);
```
Comment (sfernandez11) [inline wbooks/src/app/screens/BookList/index.tsx:18]: You can use `const books = useSelector<AppState, Book[]>(state => state.bookReducer.books);`

File: wbooks/src/redux/book/reducer.ts [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+    case actions.GET_BOOKS:
+      return {
+        ...state,
+        booksLoading: true
+      };
+
+    case actions.GET_BOOKS_SUCCESS:
+      return {
+        ...state,
+        books: action.payload
```
Comment (sfernandez11) [inline wbooks/src/redux/book/reducer.ts:27]: You forgot to set the loading in false and error in null

File: wbooks/src/redux/book/reducer.ts [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+      };
+
+    case actions.GET_BOOKS_FAILURE:
+      return {
+        ...state,
+        bookError: action.payload
```
Comment (sfernandez11) [inline wbooks/src/redux/book/reducer.ts:33]: Forgot put loding in false

**Review by sfernandez11 — COMMENTED** [review]
Last comments...

File: wbooks/src/redux/book/reducer.ts (hunk 12 above)
Comment (nathan-hyan) [reply wbooks/src/redux/book/reducer.ts:27]: Is it really necessary? I mean, it already pulls it from the initial state

Comment (nathan-hyan) [reply wbooks/src/redux/book/reducer.ts:27]: and with every run, the state gets initially assigned to initialState

**Review by sfernandez11 — APPROVED** [review]
🥇

**Pull request merged**